File: tests/bindings/smoke.py
"""Release smoke test for an installed `shacl2cypher` Python package.

Usage:
    python tests/bindings/smoke.py <s2c-fixture>   # wheel with both backends
    python tests/bindings/smoke.py --compile-only  # source build without backends

Run it outside `crates/s2c-python` so the installed package is imported.
"""

import subprocess
import sys
import tempfile
from pathlib import Path

import shacl2cypher

CORE = Path(__file__).resolve().parents[2] / "tests" / "conformance" / "core"


def main(args: list) -> None:
    compile_only = args == ["--compile-only"]
    expected = [] if compile_only else ["neo4j", "ladybug"]
    backends = shacl2cypher.available_backends()
    assert backends == expected, f"backends {backends}, expected {expected}"

    compilation = shacl2cypher.compile([CORE / "in-has-value.ttl"], dialect="neo4j", node_key="id")
    assert compilation.manifest["compilerVersion"] == shacl2cypher.__version__
    assert "// name: PersonShape.status.in" in compilation.cypher

    if not compile_only:
        (fixture_tool,) = args
        with tempfile.TemporaryDirectory() as directory:
            database = Path(directory) / "graph.lbug"
            subprocess.run(
                [fixture_tool, "ladybug-db", str(CORE / "in-has-value.yaml"), str(database)],
                check=True,
            )
            with shacl2cypher.Ladybug(database) as db:
                report = db.validate([CORE / "in-has-value.ttl"], node_key="id")
            assert report.summary["violations"] == 4, report.summary
    print(f"shacl2cypher {shacl2cypher.__version__} ok (backends: {backends})")


if __name__ == "__main__":
    main(sys.argv[1:])

File: tests/bindings/smoke.js
'use strict';

// Release smoke test for a built Node addon.
// Usage: node tests/bindings/smoke.js <package directory> <s2c-fixture>

const assert = require('node:assert/strict');
const { execFileSync } = require('node:child_process');
const fs = require('node:fs');
const os = require('node:os');
const path = require('node:path');

const CORE = path.resolve(__dirname, '..', 'conformance', 'core');

async function main([packageDir, fixtureTool]) {
  const s2c = require(path.resolve(packageDir));
  assert.deepEqual(s2c.availableBackends(), ['neo4j', 'ladybug']);

  const compilation = await s2c.compile({ shapes: [path.join(CORE, 'in-has-value.ttl')], dialect: 'neo4j', nodeKey: 'id' });
  assert.equal(compilation.manifest.compilerVersion, s2c.version);
  assert.match(compilation.cypher, /\/\/ name: PersonShape\.status\.in/);

  const database = path.join(fs.mkdtempSync(path.join(os.tmpdir(), 's2c-smoke-')), 'graph.lbug');
  execFileSync(fixtureTool, ['ladybug-db', path.join(CORE, 'in-has-value.yaml'), database]);
  const db = await s2c.Ladybug.open(database);
  try {
    const report = await db.validate({ shapes: [path.join(CORE, 'in-has-value.ttl')], nodeKey: 'id' });
    assert.equal(report.summary.violations, 4);
  } finally {
    await db.close();
  }
  console.log(`shacl2cypher ${s2c.version} ok (backends: ${s2c.availableBackends().join(', ')})`);
}

main(process.argv.slice(2)).catch((error) => {
  console.error(error);
  process.exit(1);
});
